components/script/dom/transformstream.rs

```diff
@@ -0,0 +1,909 @@
+
+        // Perform ! TransformStreamDefaultControllerClearAlgorithms(controller).
+        controller.clear_algorithms();
+
+        // React to cancelPromise:
+        let fulfillment_handler = Box::new(SourceWritableCancelPromiseResolutionHandler {
+            writable: Dom::from_ref(&writable),
+            stream: Dom::from_ref(self),
+            controller: Dom::from_ref(&controller),
+        });
+        let rejection_handler = Box::new(SourceWritableCancelPromiseRejectionHandler {
+            writable: Dom::from_ref(&writable),
+            stream: Dom::from_ref(self),
+            controller: Dom::from_ref(&controller),
+        });
+        let handler = PromiseNativeHandler::new(
+            global,
+            Some(fulfillment_handler),
+            Some(rejection_handler),
+            can_gc,
+        );
+        let realm = enter_realm(global);
+        let comp = InRealm::Entered(&realm);
+        cancel_promise.append_native_handler(&handler, comp, can_gc);
+        // Return controller.[[finishPromise]].
+        controller.get_finish_promise().ok_or(Error::Type(
+            "TransformStreamDefaultController finish_promise is null".to_string(),
+        ))
+    }
+
+    /// <https://streams.spec.whatwg.org/#transform-stream-error-writable-and-unblock-write>
+    pub(crate) fn error_writable_and_unblock_write(
+        &self,
+        global: &GlobalScope,
+        cx: SafeJSContext,
+        can_gc: CanGc,
+        reason: js::gc::HandleValue,
+    ) -> Fallible<()> {
+        // Perform ! TransformStreamDefaultControllerClearAlgorithms(stream.[[controller]]).
+        self.controller
+            .get()
+            .ok_or(Error::Type(
+                "TransformStreamDefaultController stream is null".to_string(),
+            ))?
+            .clear_algorithms();
+
+        // Perform ! WritableStreamDefaultControllerErrorIfNeeded(stream.[[writable]].[[controller]], e).
+        let stream = self.writable.get().ok_or(Error::Type(
+            "TransformStream writable stream is null".to_string(),
+        ))?;
+        stream
+            .get_default_controller()
+            .error_if_needed(cx, reason, global, can_gc);
+
+        // Perform ! TransformStreamUnblockWrite(stream).
+        self.unblock_write(global, can_gc)
+    }
+
+    /// <https://streams.spec.whatwg.org/#transform-stream-unblock-write>
+    pub(crate) fn unblock_write(&self, global: &GlobalScope, can_gc: CanGc) -> Fallible<()> {
+        // If stream.[[backpressure]] is true, perform ! TransformStreamSetBackpressure(stream, false).
+        if self.backpressure.get().unwrap_or(false) {
+            self.set_backpressure(false, global, can_gc)?;
+        }
+
+        Ok(())
+    }
+
+    /// <https://streams.spec.whatwg.org/#transform-stream-set-backpressure>
+    pub(crate) fn set_backpressure(
+        &self,
+        backpressure: bool,
+        global: &GlobalScope,
```

Comment:
I usually put global before any algo specific argument(and context, if present, as first).